### PR TITLE
e2e-distinct plugin Test Scenarios

### PR DESCRIPTION
**Pull request by itsmekumari**
@sawantpritam Please review.

### CODE REVIEW

File: e2e-test/src/test/features/BigQueryToDistinct.feature [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    Then Enter input plugin property: "referenceName" with value: "BQReferenceName"
+    Then Enter input plugin property: "dataset" with value: "dataset"
+    Then Enter input plugin property: "table" with value: "bqSourceTable"
+    Then Click on the Get Schema button
+    Then Validate "BigQuery" plugin properties
+    Then Close the Plugin Properties page
+    Then Navigate to the properties page of plugin: "Distinct"
+    Then Enter Distinct plugin Fields "distinctGCSValidFields"
+    Then Enter input plugin property: "numberOfPartitions" with value: "distinctValidPartitions"
+    Then Click on the Get Schema button
```
Comment (sawantpritam) [inline e2e-test/src/test/features/BigQueryToDistinct.feature]: Validate expected schema

File: e2e-test/src/test/features/BigQueryToDistinct.feature [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+    Then Verify the preview run status of pipeline in the logs is "succeeded"
+    Then Close the pipeline logs
+    Then Close the preview
+    Then Deploy the pipeline
+    Then Run the Pipeline in Runtime
+    Then Wait till pipeline is in running state
+    Then Open and capture logs
+    Then Verify the pipeline status is "Succeeded"
+    Then Close the pipeline logs
+    Then Validate OUT record count is equal to records transferred to target BigQuery table
```
Comment (sawantpritam) [inline e2e-test/src/test/features/BigQueryToDistinct.feature]: Validation needs to be distinct plugin specific

File: e2e-test/src/test/features/DistinctErrorScenarios.feature [hunk 3]
```diff
@@ -0,0 +1,32 @@
+@Distinct
+Feature: Distinct - Verify Distinct plugin error Scenarios
```
Comment (sawantpritam) [inline e2e-test/src/test/features/DistinctErrorScenarios.feature]: Check all feature descriptions.. add same prefix

`Distinct analytics -`

File: e2e-test/src/test/features/FileToDistinct.feature [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    Then Enter input plugin property: "path" with value: "csvAllDataTypeFile"
+    Then Select dropdown plugin property: "format" with option value: "csv"
+    Then Click plugin property: "skipHeader"
+    Then Click on the Get Schema button
+    Then Verify the Output Schema matches the Expected Schema: "csvAllDataTypeFileSchema"
+    Then Validate "File" plugin properties
+    Then Close the Plugin Properties page
+    Then Navigate to the properties page of plugin: "Distinct"
+    Then Enter Distinct plugin Fields "distinctBQValidFields"
+    Then Click on the Get Schema button
```
Comment (sawantpritam) [inline e2e-test/src/test/features/FileToDistinct.feature]: Verify expected schema

File: e2e-test/src/test/features/GCSToDistinct.feature [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    Then Verify the preview run status of pipeline in the logs is "succeeded"
+    Then Close the pipeline logs
+    Then Close the preview
+    Then Deploy the pipeline
+    Then Run the Pipeline in Runtime
+    Then Wait till pipeline is in running state
+    Then Open and capture logs
+    Then Verify the pipeline status is "Succeeded"
+    Then Close the pipeline logs
+    Then Validate OUT record count is equal to records transferred to target BigQuery table
```
Comment (sawantpritam) [inline e2e-test/src/test/features/GCSToDistinct.feature]: Validation needs to be distinct plugin specific

File: e2e-test/src/test/java/io/cdap/plugin/distinct/stepsdesign/DistinctAnalytics.java [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+import io.cucumber.java.en.Then;
+import org.junit.Assert;
+import stepsdesign.BeforeActions;
+
+import java.io.IOException;
+
+/**
+ * Distinct Plugin related step design.
+ */
+public class DistinctAnalytics implements CdfHelper {
```
Comment (sawantpritam) [inline e2e-test/src/test/java/io/cdap/plugin/distinct/stepsdesign/DistinctAnalytics.java]: Remove` implements CdfHelper`

File: e2e-test/src/test/java/io/cdap/plugin/utils/E2ETestConstants.java [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+ */
+package io.cdap.plugin.utils;
+
+/**
+ * E2E test constants.
+ */
+public class E2ETestConstants {
+  public static final String ERROR_MSG_FILE_INVALID_OUTPUTFIELD = "errorMessageFileInvalidOutputField";
+  public static final String ERROR_MSG_DISTINCT_INVALID_FIELDS = "errorMessageDistinctInvalidFields";
+  public static final String ERROR_MSG_DISTINCT_INVALID_NUMBEROFPARTITIONS = "errorMessageDistinctInvalidPartitions";
```
Comment (sawantpritam) [inline e2e-test/src/test/java/io/cdap/plugin/utils/E2ETestConstants.java]: unused

**Review by sawantpritam — CHANGES_REQUESTED** [review]

File: e2e-test/src/test/features/BigQueryToDistinct.feature (hunk 1 above)
Comment (itsmekumari) [reply e2e-test/src/test/features/BigQueryToDistinct.feature]: Added step

File: e2e-test/src/test/features/BigQueryToDistinct.feature (hunk 2 above)
Comment (itsmekumari) [reply e2e-test/src/test/features/BigQueryToDistinct.feature]: Changed step

File: e2e-test/src/test/features/DistinctErrorScenarios.feature (hunk 3 above)
Comment (itsmekumari) [reply e2e-test/src/test/features/DistinctErrorScenarios.feature]: Change done

File: e2e-test/src/test/features/FileToDistinct.feature (hunk 4 above)
Comment (itsmekumari) [reply e2e-test/src/test/features/FileToDistinct.feature]: Added step

File: e2e-test/src/test/features/GCSToDistinct.feature (hunk 5 above)
Comment (itsmekumari) [reply e2e-test/src/test/features/GCSToDistinct.feature]: Changed step

File: e2e-test/src/test/java/io/cdap/plugin/distinct/stepsdesign/DistinctAnalytics.java (hunk 6 above)
Comment (itsmekumari) [reply e2e-test/src/test/java/io/cdap/plugin/distinct/stepsdesign/DistinctAnalytics.java]: As discussed, not removed.

File: e2e-test/src/test/java/io/cdap/plugin/utils/E2ETestConstants.java (hunk 7 above)
Comment (itsmekumari) [reply e2e-test/src/test/java/io/cdap/plugin/utils/E2ETestConstants.java]: Removed unused

File: e2e-test/src/test/features/BigQueryToDistinct.feature [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+    Then Verify the preview run status of pipeline in the logs is "succeeded"
+    Then Close the pipeline logs
+    Then Close the preview
+    Then Deploy the pipeline
+    Then Run the Pipeline in Runtime
+    Then Wait till pipeline is in running state
+    Then Open and capture logs
+    Then Verify the pipeline status is "Succeeded"
+    Then Close the pipeline logs
+    Then Validate OUT record count of distinct is equal to IN record count of sink
```
Comment (sawantpritam) [inline e2e-test/src/test/features/BigQueryToDistinct.feature]: Then Validate target table "bqTargetTable" records count is equal to source table "bqSourceTable" records count with distinct fields "distinctBQFieldsList"

implementation -> fetch the record count with distinct query from source table and assert with total records count of target table

**Review by sawantpritam — CHANGES_REQUESTED** [review]

File: e2e-test/src/test/features/BigQueryToDistinct.feature (hunk 8 above)
Comment (itsmekumari) [reply e2e-test/src/test/features/BigQueryToDistinct.feature]: Added the validation step

File: e2e-test/src/test/java/io/cdap/plugin/distinct/stepsdesign/DistinctAnalytics.java [hunk 9]
```diff
@@ -66,4 +69,27 @@ public void verifyDistinctPluginInLineErrorMessageForIncorrectFields(String fiel
...
+  public void validateTargetTableRecordsCountIsEqualToSourceTableRecordsCountWithDistinctFields(String targetTable,
+                                                                                 String sourceTable, String fields)
+    throws IOException, InterruptedException {
+      Optional<String> result = BigQueryClient
+        .getSoleQueryResult("SELECT COUNT(*) FROM (SELECT  distinct "  +  PluginPropertyUtils.pluginProp(fields) +
+                              " as records FROM `"
+      + (PluginPropertyUtils.pluginProp("projectId")) + "."
+      + (PluginPropertyUtils.pluginProp("dataset")) + "."
+      + PluginPropertyUtils.pluginProp(sourceTable) + "`)");
+    int distinctCountBQTable = 0;
```
Comment (sawantpritam) [inline e2e-test/src/test/java/io/cdap/plugin/distinct/stepsdesign/DistinctAnalytics.java]: line 84 to 89 can be replaced with single line ->

`int distinctCountBQTable = result.map(Integer::parseInt).orElse(0);`

**Review by sawantpritam — CHANGES_REQUESTED** [review]

File: e2e-test/src/test/java/io/cdap/plugin/distinct/stepsdesign/DistinctAnalytics.java (hunk 9 above)
Comment (itsmekumari) [reply e2e-test/src/test/java/io/cdap/plugin/distinct/stepsdesign/DistinctAnalytics.java]: Change done

**Pull request closed without merging**